### PR TITLE
Add fill-extrusion-vertical-gradient property

### PR DESCRIPTION
**Pull request by lbud**
This PR adds a new boolean data-constant paint property, `fill-extrusion-vertical-gradient` ( 🤔 a bit verbose — I'm open to better ideas). It defaults to `true` (the current behavior), but if set to `false`, does not apply a vertical gradient along the sides of fill-extrusion features.

Closes https://github.com/mapbox/mapbox-gl-js/issues/5768.

I only ran benchmarks for `LayerFillExtrusion` since nothing else should be affected, and it shows no real change: https://bl.ocks.org/lbud/raw/d353b63e34c7fcb6e4ec2698370fd6b9/

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs (docs page update will be generated in release)
 - [x] post benchmark scores
 - [x] manually test the debug page
 - [x] tag @mapbox/studio and/or `@mapbox/maps-design` if this PR includes style spec changes

### CODE REVIEW

**Comment by ryanhamley** [comment]
Is @mapbox/studio ok with including this? Seems like this is a desired option.

**Comment by elifitch** [comment]
In general we're a little concerned about the spec getting bloated, but we're good with this addition; it'll be easy enough for us to support

**Comment by ryanhamley** [comment]
I rebased this PR against `master` and updated it to use uniform bindings. I think it's good to go; the test results look good.

screen shot 2018-09-24 at 2 13 21 pm

File: src/render/program/fill_extrusion_program.js [hunk 1]
```diff
@@ -24,7 +24,8 @@ export type FillExtrusionUniformsType = {|
     'u_matrix': UniformMatrix4f,
     'u_lightpos': Uniform3f,
     'u_lightintensity': Uniform1f,
-    'u_lightcolor': Uniform3f
+    'u_lightcolor': Uniform3f,
+    'u_vertical_gradient': Uniform1f
```
Comment (mollymerp) [inline src/render/program/fill_extrusion_program.js]: this is just 1 or 0 right? in that case we can use an integer uniform `Uniform1i` to follow the convention of other uniforms used as booleans such as `u_is_zoom_constant` for symbol icons/sdfs

File: src/render/program/fill_extrusion_program.js [hunk 2]
```diff
@@ -80,7 +84,8 @@ const extrusionTextureUniforms = (context: Context, locations: UniformLocations)
 
 const fillExtrusionUniformValues = (
     matrix: Float32Array,
-    painter: Painter
+    painter: Painter,
+    verticalGradient: number
```
Comment (mollymerp) [inline src/render/program/fill_extrusion_program.js]: For clarity, I lean towards passing this value to the `*UniformValues` functions as a boolean, and casting only as the value is bound to its uniform location.

File: src/shaders/fill_extrusion.vertex.glsl [hunk 3]
```diff
@@ -47,7 +48,11 @@ void main() {
 
     // Add gradient along z axis of side surfaces
     if (normal.y != 0.0) {
-        directional *= clamp((t + base) * pow(height / 150.0, 0.5), mix(0.7, 0.98, 1.0 - u_lightintensity), 1.0);
+        // This avoids another branching statement, but multiplies by a constant of 0.84 if no vertical gradient,
```
Comment (mollymerp) [inline src/shaders/fill_extrusion.vertex.glsl]: do we have any idea where the magic -0.84 comes from? maybe nicki would know

Comment (ryanhamley) [reply src/shaders/fill_extrusion.vertex.glsl]: why would Nicki know? i have no idea where it comes from

File: src/render/program/fill_extrusion_program.js (hunk 2 above)
Comment (ryanhamley) [reply src/render/program/fill_extrusion_program.js]: `Uniform1i` requires a number input so casting the boolean to a number only when binding to the location causes a Flow error. Is there a way around that?

File: src/shaders/fill_extrusion.vertex.glsl (hunk 3 above)
Comment (mollymerp) [reply src/shaders/fill_extrusion.vertex.glsl]: Nicki worked on the extrusion shaders w lbud

Comment (ryanhamley) [reply src/shaders/fill_extrusion.vertex.glsl]: @nickidlugash do you remember where this magic number came from?

File: src/render/program/fill_extrusion_program.js (hunk 2 above)
Comment (mollymerp) [reply src/render/program/fill_extrusion_program.js]: yep you use the `+` like here https://github.com/mapbox/mapbox-gl-js/blob/master/src/render/program/symbol_program.js#L116

Comment (ryanhamley) [reply src/render/program/fill_extrusion_program.js]: Ah yeah, I was mistakenly casting to a number elsewhere in the program, causing a flow error. This is fixed now.

File: src/render/program/fill_extrusion_program.js (hunk 1 above)
Comment (ryanhamley) [reply src/render/program/fill_extrusion_program.js]: This actually needs to be a float because the calculation in the shader has float constants.

**Comment by ryanhamley** [comment]
I updated the way the gradient is calculated (removed the 0.84 constant) to allow the full spectrum of light intensity to be used in determining the final color.

screen shot 2018-09-25 at 2 49 30 pm

**Pull request merged**